dataclients/kubernetes/traffic.go

```diff
@@ -0,0 +1,122 @@
+package kubernetes
+
+import (
+	"github.com/zalando/skipper/dataclients/kubernetes/definitions"
+	"github.com/zalando/skipper/eskip"
+	"github.com/zalando/skipper/predicates"
+)
+
+// backendTraffic specifies whether a given backend is allowed to receive any traffic and
+// modifies route to receive the desired traffic portion
+type backendTraffic interface {
+	allowed() bool
+	apply(*eskip.Route)
+}
+
+// GetBackendTrafficCalculator returns a function that calculates backendTraffic for each backend using specified algorithm
+func GetBackendTrafficCalculator[T definitions.WeightedBackend](algorithm string) func(b []T) map[string]backendTraffic {
```

Comment:
This factory function is generic to support both ingress and routegroup weighted backends and avoid copying slices (see https://go.dev/doc/faq#convert_slice_of_interface and https://stackoverflow.com/questions/12753805/type-converting-slices-of-interfaces)

Alternatively it could use RouteGroup's `*BackendReference` but:
* it is defined for routegroup but would be used by ingress as well
* `BackendReference.Weight` int type needs to be changed to float64 along with additional validation that user-provided values are whole numbers